Move 'bin/worker.php' to 'bin/console.php worker'

nupplaphil opened the pull request:
- [x] Move `App` logic into console
- [x] Refactor execution of worker in the `Worker` and `Cron` class (currently, they spawn per `worker.php` exection)

nupplaphil committed:
Move 'bin/worker.php' to 'bin/console.php worker'

nupplaphil committed:
Update doc

MrPetovan commented on src/Console/Worker.php:
```diff
@@ -57,11 +57,11 @@ protected function getHelp(): string
-	bin/console daemon start -f
-		Starts the daemon in the foreground
+	bin/console worker -n
+		Starts the worker without executing the cronob
```
Besides the typo, what does this mean to not execute the cronjob?

nupplaphil replied:
It doesn't execute https://github.com/friendica/friendica/blob/develop/src/Worker/Cron.php#L22 , whatever impact that has :-) 

maybe @annando can explain `no_cron` and `spawn` in a more precise way and I can take it for the help-texte

annando replied:
Normally the system will automatically execute a call to the `cron` tasks on startup. With this parameter it doesn't do so.

nupplaphil replied:
And why is it possible? What benefit do we get?

And is `spawn` correctly a parameter to spawn **another** worker process beside the current one?

MrPetovan replied:
Then it should read something like this:
```suggestion
		Starts the worker without executing other recurring tasks
```

nupplaphil committed:
Replace 'bin/daemon.php' and 'bin/jetstream.php' calls and help

nupplaphil committed:
Replace 'bin/worker.php' calls and help

nupplaphil committed:
Adapt run-command for console.php executions

nupplaphil commented on src/App.php:
```diff
@@ -215,53 +213,6 @@ public function processConsole(array $argv): void
-	{
-		$this->setupContainerForAddons();
-
-		$this->setupContainerForLogger(LogChannel::WORKER);
```
@art4 - with this refactoring, we loose all different logging channels. It's now `console` all the time :-/

One "quick hack" would be to distinguish at `processConsole`, based on the subcommand, like
`if in_array($argv, 'daemon'] { LogChannel::DAEMON }`
but it feels ultra-hacky!

Do you have another idea? The problem is that we setup `DI::` afterwards and create some classes, which then will use the `LoggerInterface`, which makes it hacky as well to switch the channel afterwards. In `Monolog`, it's not even switchable (part of the constructor)

Art4 replied:
I'm afk atm. Solution could be named instances.
https://r.je/dice#example3-6

nupplaphil committed:
Make PHP-CS happy

nupplaphil committed:
Move Container logic into own class

nupplaphil committed:
Fixup LogChannels

nupplaphil committed:
Update src/Console/Worker.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

nupplaphil commented:
Fixed the Log channel channel issue :-)

I moved the container setup into the class `Container` and split `App` and `Console`. So I broke up the circle dependency and can now set the LogChannel before using the container the first time :-)

nupplaphil committed:
Add doc & fix test

nupplaphil committed:
Add license

nupplaphil committed:
Add UnitTest for Container

nupplaphil committed:
Make PHP-CS happy :-)

nupplaphil committed:
Use correct typing

nupplaphil committed:
Make PHP-CS happy

nupplaphil committed:
Fix default log channel

nupplaphil committed:
Centralize deprecated check

MrPetovan commented on bin/worker.php:
```diff
@@ -7,6 +7,8 @@
  *
  * Starts the background processing
+ *
+ * @deprecated 2025.01 use bin/console.php worker instead
```
```suggestion
 * @deprecated 2025.02 use bin/console.php worker instead
```

MrPetovan commented on src/Console/AbstractConsole.php:
```diff
@@ -0,0 +1,38 @@
+	 */
+	protected function checkDeprecated(string $command): void
+	{
+		if (substr($this->executable, -strlen('bin/console.php')) !== 'bin/console.php') {
```
Can you please move the `bin/console.php` string to a constant?

nupplaphil committed:
Update deprecated version

nupplaphil committed:
Add cons

Art4 commented:
Sorry for being late at the review. This PR unfortunately reverted many of the improvements that came with the centralization in the App class. The new Container class is great but does too much, like 
- `setupContainerForAddons()`
- `setupContainerForLogger()`
- `setupLegacyServiceLocator()` 
- and `registerErrorHandler()`

These calls are no tasks for the Containet and should be in the App class where only `Container::addRule()` is called.

The Container has also duplicated code from the App class, so now we have again 2 places for `registerTemplateEngine()`.

The workflow should  always be:

1. index.php/console.php is called
2. Container is created
3. App is created by Container
4. App modifies Container for addons
5. App setup Logger
6. App setup DI class
7. App registers Error handler
8. App registers template engine if needed
9. App calls Module/Console class

This way we can control every request or CLI command through the App class.

Can I create a PR to refactor the Container and App class?

Art4 commented on src/Console/AbstractConsole.php:
```diff
@@ -0,0 +1,39 @@
+	 */
+	protected function checkDeprecated(string $command): void
+	{
+		if (substr($this->executable, -strlen(CoreConsole::getDefaultExecutable())) === CoreConsole::getDefaultExecutable()) {
```
The condition is wrong. Using `bin/conlose.php daemon` returns the message `'bin/console.php' is deprecated and will removed. Please use 'bin/console.php daemon' instead`

```suggestion
		if (substr($this->executable, -strlen(CoreConsole::getDefaultExecutable())) !== CoreConsole::getDefaultExecutable()) {
```

Art4 replied:
Will be fixed in #14677.
